virtual_ecosystem/models/plants/plants_model.py

```diff
@@ -353,98 +440,122 @@ def estimate_gpp(self, time_index: int) -> None:
         The GPP for each cohort is then estimated by mutiplying the cohort canopy area
         within each layer by GPP and the time elapsed in seconds since the last update.
 
-        Warning:
-            At present this method checks that the required forcing variables exist, but
-            asserts a constant fixed light use efficiency rather than using the P Model.
+        .. TODO:
+
+            * This function populates evapotranspiration but the calculation is
+              currently only estimating _transpiration_
+              `#704 <https://github.com/ImperialCollegeLondon/virtual_ecosystem/issues/704>`_
+            * Conversion of transpiration from `µmol m-2` to `mm m-2` currently ignores
+              density changes with conditions:
+              `#723 <https://github.com/ImperialCollegeLondon/virtual_ecosystem/issues/723>`_
 
         Args:
             time_index: The index along the time axis of the forcing data giving the
                 time step to be used to estimate GPP.
 
         Raises:
             ValueError: if any of the P Model forcing variables are not defined.
-        """
 
-        # For the moment check the data presence and dimensionality
-        # 1 ) PModelEnvironment vars
-        forcing_vars = (
-            "air_temperature",
-            "vapour_pressure_deficit",
-            "atmospheric_pressure",
-            "atmospheric_co2",
-        )
 
-        # Loop over the P Model environment forcing variables, checking they are found
-        for var in forcing_vars:
-            if var not in self.data:
-                msg = f"Variable missing from estimate_gpp: {var}"
-                LOGGER.critical(msg)
-                raise ValueError(msg)
-
-        # # Next step will to calculate the PModelEnvironment and fit the PModel
-        # pmodel_env = PModelEnvironment(
-        #     tc = self.data["air_temperature"].isel(time=time_index).data
-        #     vpd = self.data["vapour_pressure_deficit"].isel(time=time_index).data
-        #     patm = self.data["atmospheric_pressure"].isel(time=time_index).data
-        #     co2 = self.data["atmospheric_co2"].isel(time=time_index).data
-        #     const = self.pmodel_consts
-        # )
-        # pmodel = PModel(pmodel_env)
-        #
-        # This will give an array of the light use efficiency per layer per cell,
+        """
 
-        # Get an array where populated canopy layers are one otherwise nan
-        filled_canopy = xr.where(
-            (self.data["layer_heights"] * self._canopy_layer_indices[:, None]) > 0,
-            1,
-            np.nan,
+        # Estimate the light use efficiency of leaves within each canopy layer within
+        # each grid cell. The LUE is set purely by the environmental conditions, which
+        # are shared across cohorts so we can calculate all layers in all cells.
+        pmodel_env = PModelEnvironment(
```

Comment:
Note that these arrays contain all the layers, including above, soil, and nan canopy layers. Would it make sense to select by filled canopy index?

Reply:
I did consider it - I'm not sure whether the subsetting and reinsertion is actually a net benefit over sticking oversized arrays through `pyrealm`, so I kept it simple!